### PR TITLE
HPCC-22956 Turn on -Wall -Wextra for Thor

### PR DESCRIPTION
**Pull request by jakesmith**
Signed-off-by: Jake Smith <jake.smith@lexisnexisrisk.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [x] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

File: thorlcr/activities/lookupjoin/thlookupjoinslave.cpp [hunk 1]
```diff
@@ -968,9 +968,9 @@ class CInMemJoinBase : public CSlaveActivity, public CAllOrLookupHelper<HELPER>,
     }
     inline bool incNotifyCountAndCheck()
     {
-        if (atomic_add_and_read(&interChannelToNotifyCount, 1) == queryJob().queryJobChannels())
+        if (++interChannelToNotifyCount == queryJob().queryJobChannels())
```
Comment (jakesmith) [inline thorlcr/activities/lookupjoin/thlookupjoinslave.cpp]: @ghalliday - please double check this.

**Comment by jakesmith** [comment]
@ghalliday - please review

File: thorlcr/activities/result/thresult.cpp [hunk 2]
```diff
@@ -77,7 +77,6 @@ class CResultActivityMaster : public CMasterActivity
         if (!results && !abortSoon)
         {
             ActPrintLog("WARNING: no results");
-            IHThorRemoteResultArg *helper = (IHThorRemoteResultArg *)queryHelper();
```
Comment (richardkchapman) [inline thorlcr/activities/result/thresult.cpp]: Makes the comment on the line below a bit less meaningful.

File: thorlcr/activities/thdiskbase.cpp [hunk 3]
```diff
@@ -319,7 +318,6 @@ void CWriteMasterBase::publish()
                     partDesc->queryProperties().setPropInt64("@size", 0);
                     p++;
                 }
-                clusterIdx++;
```
Comment (richardkchapman) [inline thorlcr/activities/thdiskbase.cpp]: This looks like a bugfix rather than a pure warning removal, and should perhaps be in a separate jira?

Comment (jakesmith) [reply thorlcr/activities/thdiskbase.cpp]: yes, I suspect it's [almost] benign, in that multi cluster target files on a remote target that's a different width to the local cluster (needs to be larger) are probably something that has never been used.
I'll remove and separate.

File: thorlcr/activities/result/thresult.cpp (hunk 2 above)
Comment (jakesmith) [reply thorlcr/activities/result/thresult.cpp]: yes, the exception is good enough, I'll remove the comments.

File: thorlcr/master/thmastermain.cpp [hunk 4]
```diff
@@ -461,7 +460,7 @@ class CThorEndHandler : public CSimpleInterface, implements IThreaded
     void start(unsigned timeoutSecs)
     {
         bool expected = false;
-        if (started.compare_exchange_strong(expected, true));
+        if (started.compare_exchange_strong(expected, true))
```
Comment (ghalliday) [inline thorlcr/master/thmastermain.cpp]: this looks like a bug fix.  Should it also be in a separate jira/PR? same for the line below.

**Review by ghalliday — COMMENTED** [review]
Looks like a worthwhile change.  One similar comment to Richard's.

**Comment by ghalliday** [comment]
@jakesmith

File: thorlcr/master/thmastermain.cpp (hunk 4 above)
Comment (jakesmith) [reply thorlcr/master/thmastermain.cpp]: it's in another PR already, but wasn't upmerged (but is now), will disappear when I rebase.

Comment (jakesmith) [reply thorlcr/master/thmastermain.cpp]: @ghalliday now rebased

**Comment by jakesmith** [comment]
@ghalliday - addressed comments, ready to merge.

File: thorlcr/activities/lookupjoin/thlookupjoinslave.cpp (hunk 1 above)
Comment (ghalliday) [reply thorlcr/activities/lookupjoin/thlookupjoinslave.cpp]: I agree it is the same.

**Pull request merged**